File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
from collections import defaultdict

def solution(friends, gifts):
    giftnum = defaultdict(int)
    total = defaultdict(lambda: defaultdict(int))
    
    for g in gifts:
        give, receive = map(str, g.split())
        total[give][receive] += 1
        giftnum[give] += 1
        giftnum[receive] -= 1
        
    received = {name: 0 for name in friends}
    n = len(friends)
    for i in range(n):
        for j in range(i+1, n):
            a,b = friends[i],friends[j]
            atob = total.get(a, {}).get(b, 0)
            btoa = total.get(b, {}).get(a, 0)
            
            if atob > btoa: #준게 더 많음
                received[a] += 1
            elif atob < btoa:
                received[b] += 1
            else: #같으면선물지수로
                if giftnum[a] > giftnum[b]:
                    received[a] += 1
                elif giftnum[b] > giftnum[a]:
                    received[b] += 1
    return max(received.values()) if received else 0

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <sstream>
#include <iostream>
#include <algorithm>

using namespace std;

int solution(vector<string> friends, vector<string> gifts) {
    int answer = 0;
    unordered_map<string, int> total;
    
    unordered_map<string, unordered_map<string, int>> gift;
    unordered_map<string, int> giftnum;
    for(auto g: gifts) {
        stringstream ss(g);
        string a,b;
        ss >> a >> b;
        gift[a][b] += 1;
        giftnum[a] += 1;
        giftnum[b] -= 1;
    }
    
    for(int i=0;i<friends.size();i++) {
        for(int j=i+1;j<friends.size();j++) {
            string a = friends[i];
            string b = friends[j];
            int atob = gift[a][b];
            int btoa = gift[b][a];
            
            if (atob > btoa) total[a]++;
            else if (atob < btoa) total[b]++;
            else {
                int numa = giftnum[a];
                int numb = giftnum[b];
                if (numa > numb) total[a]++;
                else if (numa < numb) total[b]++;
            }
        }
    }
    
    for(auto [k,v]: total) {
        answer = max(answer,v);
    }
    
    return answer;
}